spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java

```diff
@@ -51,8 +55,10 @@
 	private String dispatcherServletPath = "/";
 	private String zuulServletPath;
 
-	private AtomicReference<Map<String, ZuulRoute>> routes = new AtomicReference<>();
+	private AtomicReference<Map<String, List<ZuulRoute>>> routes = new AtomicReference<>();
```

Comment:
this is to do with how `locateRoutes` method works and creates the map.

Old implementation
```java
	protected Map<String, ZuulRoute> locateRoutes() {
		LinkedHashMap<String, ZuulRoute> routesMap = new LinkedHashMap<String, ZuulRoute>();
		for (ZuulRoute route : this.properties.getRoutes().values()) {
			routesMap.put(route.getPath(), route);
		}
		return routesMap;
	}
```
Since this is using path as the key, having multiple routes of the same path overwrites the value for the key.
```yaml
zuul:
  routes:
    users-read:
      path: /myusers/service
      url: http://user-service/readusers/newservice
      methods: GET
    users-write:
      path: /myusers/service
      url: http://user-service/readusers/newservice
      methods: PUT, POST
```
The will end up as a single route for key `/myusers/service`, so had to create a list to have multiple routes against the key `/myusers/service`. The `locateRoute` method is called by `DiscoveryClientRouteLocator` as well so creating a list seemed less dangerous than making an overall change for `locateRoutes`.

But now that I think about it, this may not be the ideal solution. We'll get `404 Not Found` if path + method is not found e.g. in the above yaml `PATCH` will give `404` while ideally it should return `405 Method Not Allowed` since resource exists but method is not available.

The above leads me to think that it should be something like below
```yaml
zuul:
  routes:
    users:
      path: /myusers/service
      url: http://default-user-service
      allowed-methods: GET, PUT, POST, PATCH
      - methods: GET
        url: http://read-user-service
      - methods: PUT, POST
        url: http://write-user-service
```
* If method is not in the list of `allowed-methods` e.g. `DELETE` in this case, then Zuul should throw `405 Method not Allowed`. 
* If method is allowed but not specified explicitly e.g. `PATCH` in this case, then default url should take effect.
* And of coarse, if nothing is specified then allow all methods to default url.

I might have taken it too far but this will really make it shine. Let me know what you think about it?